src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/StringCache.cs

```diff
@@ -0,0 +1,162 @@
+        {
+            lock (_lock)
+            {
+                if (key is null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                if (TryGetValue(key, out var result))
+                {
+                    return result!;
+                }
+                else
+                {
+                    var weakRef = new Entry(key);
+                    _hashSet.Add(weakRef);
+
+                    // Whenever we expand lets clean up dead references
+                    if (_capacity <= _hashSet.Count)
+                    {
+                        Cleanup();
+                        _capacity *= 2;
+                    }
+
+                    return key;
+                }
+            }
+        }
+
+        private void Cleanup()
+        {
+            _hashSet.RemoveWhere((weakRef) =>
+            {
+                return !weakRef.IsAlive;
+            });
+        }
+
+        private bool TryGetValue(string key, out string? value)
+        {
+            if (_hashSet.Contains(new Entry(key)))
+            {
+                var exfiltrator = (ExfiltratingEqualityComparer)_hashSet.Comparer;
+                var val = exfiltrator.LastEqualValue!;
+                if (val.TryGetTarget(out var target))
+                {
+                    value = target!;
+                    return true;
+                }
+                else
+                {
+                    throw new InvalidOperationException("HashSet contained entry but we were unable to get the value from it. This should be impossible");
+                }
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// This is a gross hack to do a sneaky and get the value inside the HashSet out given the lack of any Get operations in netstandard2.0.
+        /// If we ever upgrade to 2.1 delete this and just use the built in TryGetValue method.
+        /// </summary>
+        /// <remarks>
+        /// This is fragile on the ordering of the values passed to the EqualityComparer by HashSet.
+        /// If that ever switches we have to react, if it becomes indeterminate we have to abandon this strategy.
+        /// </remarks>
+        private class ExfiltratingEqualityComparer : IEqualityComparer<Entry>
+        {
+            public Entry? LastEqualValue { get; private set; }
+
+            public bool Equals(Entry x, Entry y)
+            {
+                if (x.Equals(y))
+                {
+                    LastEqualValue = x;
+                    return true;
+                }
+                else
+                {
+                    LastEqualValue = null;
+                    return false;
+                }
+            }
+
+            public int GetHashCode(Entry obj)
+            {
+                return obj.TargetHashCode;
+            }
+        }
+
+        private class Entry
```

Comment:
Thoughts on using a struct here instead of a class?

Reply:
I thought this would be straight-forward savings since we can't ever change an Entry, but it turns out not to work (at least without significant re-work). The problem is once we make Entry a struct the `ReferenceEquals` comparison at the end of `Equals` [starts failing due to Value Boxing](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2013?view=vs-2019). If we can't get equality against itself then it becomes impossible to evict an entry from the HashSet (because it checks that it's equal to itself before Removing things inside HashSet), and probably other HashSet scenarios go off the rails.